File: backend/app/schemas/family.py
"""Pydantic schemas for family management."""

from datetime import datetime
from pydantic import BaseModel, Field


class FamilyCreate(BaseModel):
    """Create a new family."""
    name: str = Field(..., min_length=2, max_length=100)


class FamilyUpdate(BaseModel):
    """Update family details."""
    name: str | None = Field(None, min_length=2, max_length=100)


class FamilyRead(BaseModel):
    """Family response model."""
    id: int
    name: str
    invite_code: str | None
    invite_expires_at: datetime | None
    created_at: datetime
    is_active: bool
    member_count: int | None = None

    model_config = {"from_attributes": True}


class FamilyMemberRead(BaseModel):
    """Family member response."""
    id: int
    user_id: int
    user_name: str
    user_role: str  # parent | child
    role_in_family: str  # admin | parent | child
    joined_at: datetime

    model_config = {"from_attributes": True}


class AddChildRequest(BaseModel):
    """Add a child to the family."""
    name: str = Field(..., min_length=2, max_length=100)
    pin: str = Field(..., min_length=4, max_length=8, pattern=r"^\d+$")
    allowed_devices: list[str] | None = None


class DeviceProviderConfig(BaseModel):
    """Device provider configuration."""
    device_type: str  # phone | pc | tablet | console
    provider_type: str  # kisi | family_link | manual
    provider_settings: dict | None = None


class DeviceProviderRead(BaseModel):
    """Device provider response."""
    id: int
    device_type: str
    provider_type: str
    provider_settings: dict | None

    model_config = {"from_attributes": True}


class InviteCodeResponse(BaseModel):
    """Generated invite code."""
    invite_code: str
    expires_at: datetime


class JoinFamilyRequest(BaseModel):
    """Join a family with invite code."""
    invite_code: str = Field(..., min_length=6, max_length=16)


class ProviderInfo(BaseModel):
    """Available provider info."""
    code: str
    name: str
    description: str | None
    requires_tan_pool: bool

File: frontend/lib/models/family.dart
class Family {
  final int id;
  final String name;
  final String? inviteCode;
  final DateTime? inviteExpiresAt;
  final DateTime createdAt;
  final bool isActive;

  const Family({
    required this.id,
    required this.name,
    this.inviteCode,
    this.inviteExpiresAt,
    required this.createdAt,
    this.isActive = true,
  });

  factory Family.fromJson(Map<String, dynamic> json) {
    return Family(
      id: json['id'] as int,
      name: json['name'] as String,
      inviteCode: json['invite_code'] as String?,
      inviteExpiresAt: json['invite_expires_at'] != null
          ? DateTime.parse(json['invite_expires_at'] as String)
          : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      isActive: json['is_active'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'invite_code': inviteCode,
      'invite_expires_at': inviteExpiresAt?.toIso8601String(),
      'created_at': createdAt.toIso8601String(),
      'is_active': isActive,
    };
  }
}

class FamilyMember {
  final int id;
  final int familyId;
  final int userId;
  final String roleInFamily;
  final DateTime joinedAt;
  final String? userName;

  const FamilyMember({
    required this.id,
    required this.familyId,
    required this.userId,
    required this.roleInFamily,
    required this.joinedAt,
    this.userName,
  });

  factory FamilyMember.fromJson(Map<String, dynamic> json) {
    return FamilyMember(
      id: json['id'] as int,
      familyId: json['family_id'] as int,
      userId: json['user_id'] as int,
      roleInFamily: json['role_in_family'] as String,
      joinedAt: DateTime.parse(json['joined_at'] as String),
      userName: json['user_name'] as String?,
    );
  }
}

class DeviceProvider {
  final int id;
  final int familyId;
  final String deviceType;
  final String providerType;
  final Map<String, dynamic>? providerSettings;

  const DeviceProvider({
    required this.id,
    required this.familyId,
    required this.deviceType,
    required this.providerType,
    this.providerSettings,
  });

  factory DeviceProvider.fromJson(Map<String, dynamic> json) {
    return DeviceProvider(
      id: json['id'] as int,
      familyId: json['family_id'] as int,
      deviceType: json['device_type'] as String,
      providerType: json['provider_type'] as String,
      providerSettings: json['provider_settings'] as Map<String, dynamic>?,
    );
  }
}

class RewardProviderInfo {
  final String code;
  final String name;
  final String? description;
  final bool requiresTanPool;

  const RewardProviderInfo({
    required this.code,
    required this.name,
    this.description,
    this.requiresTanPool = false,
  });

  factory RewardProviderInfo.fromJson(Map<String, dynamic> json) {
    return RewardProviderInfo(
      code: json['code'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      requiresTanPool: json['requires_tan_pool'] as bool? ?? false,
    );
  }
}
